pennylane/_qubit_device.py

```diff
@@ -826,6 +826,21 @@ def sample(self, observable, shot_range=None, bin_size=None):
 
         return samples.reshape((bin_size, -1))
 
+    def execute_and_gradients(self, circuits):
+        res = []
+        jacs = []
+
+        for circuit in circuits:
+            # Evaluations and gradients are paired, so that
+            # we can re-use the device state for the adjoint method
+            res.append(circuit.execute(self))
+            jacs.append(self.adjoint_jacobian(circuit, use_device_state=True))
+
+        return res, jacs
+
+    def gradients(self, circuits):
+        return [self.adjoint_jacobian(circuit) for circuit in circuits]
```

Comment:
I added these to the device API for consistency and to make the interface code cleaner.

Reply:
> This is helpful, but it bakes into the parent class an assumption about supported gradient methods.

Yep, I agree. I definitely think the final design decision of these two methods in the Device is TBD.

> use a different "device" gradient method when adjoint is not supported/optimal?

Good idea. Unfortunately, we don't have any other available device methods I know of at the moment 😆 In core, it's either adjoint or nothing. But I can add validation here that checks the capabilities dictionary :+1:

Somewhat larger in scope, but I spent a lot of time thinking about this problem yesterday, and prototyped up some approaches with user-defined device methods. 

I kept coming back to the following thought - perhaps the best approach for a plugin developer is simply to fully overwrite `Device.execute_and_gradients` and `Device.gradients`?

My thought process was inspired a lot by what happened with batch execute:

- We provide a 'simple' default `Device.batch_execute`, that just has a for loop over execution.

- The Braket plugin [_already_ fully overwrites `batch_execute`](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/2883b6d67641cbc90c42292d1d5af40f3553efc7/src/braket/pennylane_plugin/braket_device.py#L325) because they need to do a lot of custom parallel/batching.

So the most flexible thing for the future seems to be (a) provide a sensible (albeit slightly stupid) default implementation, and (b) allow devices to overwrite it as needed.